### ISSUE TITLE
Advice for the Implementation of BPR for Importance Sampling

### ISSUE THREAD

**Issue by adebiasio-estilos** [massquantity/LibRecommender#203]
Hi All,
I would like to implement in LibRecommender BPR with Importance Sampling.
-  http://staff.ustc.edu.cn/~liandefu/paper/pris.pdf
I have already modified the Sampling function.
What should i do to modify the loss function of BPR?

Thanks in advance
Alvise

**Comment by adebiasio-estilos** [comment]
Moreover, taking for example in consideration BPR's current implementation:
- https://github.com/massquantity/LibRecommender/blob/master/libreco/algorithms/bpr.py

how can I print at each forward pass the various loss variables?

**Comment by massquantity** [comment]
Thanks for implementing this! I've seen the paper you mentioned roughly, and I'll assume you're using TensorFlow1. 
It seems that the main difference is they add a weight(w_l) parameter in the original BPR loss. So you can add a hyperparameter such as `importance_sampling` to let users choose whether to use it. Then you will implement the weight(w_l).

At last, in [bpr.py](https://github.com/massquantity/LibRecommender/blob/master/libreco/algorithms/bpr.py#L139), the code may look like this:
```python
if self.importance_sampling:
    self.bpr_loss = w * tf.log_sigmoid(item_diff)
else:
    self.bpr_loss = tf.log_sigmoid(item_diff)
```

**Comment by massquantity** [comment]
What do you mean by various loss variables? Training loss or model's variables?

**Comment by adebiasio-estilos** [comment]
Thank you for your help @massquantity !
The methodology you propose looks good!

**Comment by adebiasio-estilos** [comment]
> What do you mean by various loss variables? Training loss or model's variables?

Yes, I mean the TF model's variables during the graph execution.
How are you used to debug the TF1 code?

**Comment by massquantity** [comment]
For example, If I want to get the `self.bpr_loss`, run `sess.run` in the [training function](https://github.com/massquantity/LibRecommender/blob/master/libreco/training/tf_trainer.py#L356), which can get the variable and its shape.
```python
for data in data_generator(shuffle, self.batch_size):
    self.sess.run(
        self.training_op,
        feed_dict={
            self.model.user_indices: data[0],
            self.model.item_indices_pos: data[1],
            self.model.item_indices_neg: data[2],
        },
    )

    feed_dict = {
        self.model.user_indices: data[0],
        self.model.item_indices_pos: data[1],
        self.model.item_indices_neg: data[2],
    }
    print(self.sess.run(self.model.bpr_loss, feed_dict=feed_dict).shape)
    print(self.sess.run(self.model.bpr_loss, feed_dict=feed_dict))
```

If you can't get a variable from `self.model`, just add a `self.` before the variable you want to get in your implementation.

**Comment by adebiasio-estilos** [comment]
Thank you very much, it works!!!
